src/V3ReuseWideTemps.cpp

```diff
@@ -0,0 +1,246 @@
+        // with the same replacee in one statement.
+        for (const auto& replaced : m_replacedInStmt) {
+            const WidthType& key = replaced.first;
+            const IdTempMap& temps = replaced.second;
+            m_freeTemps[key].insert(temps.cbegin(), temps.cend());
+        }
+    }
+    void removeUnusedVars() {
+        for (AstVar* varp : m_removedTemps) {
+            VL_DO_DANGLING(varp->unlinkFrBack()->deleteTree(), varp);
+        }
+    }
+
+    // VISITORS
+    void visit(AstIf* nodep) override {
+        // Temp may be reused when READ was in a non-taken branch.
+        m_replacedInStmt.clear();
+        iterateAndNextConstNull(nodep->condp());
+        addFreedTemps();
+
+        m_replacedInStmt.clear();
+        iterateAndNextConstNull(nodep->thensp());
+        addFreedTemps();
+
+        m_replacedInStmt.clear();
+        iterateAndNextConstNull(nodep->elsesp());
+        addFreedTemps();
+    }
+    void visit(AstNodeStmt* nodep) override {
+        m_replacedInStmt.clear();
+        iterateChildren(nodep);
+        addFreedTemps();
+    }
+    void visit(AstVar* nodep) override {
+        if (!isWideTemp(nodep)) return;
+        UINFO(5, "Populating free temp from declaration: " << nodep << endl);
+
+        if (!isExcluded(nodep)) {
+            m_freeTemps[nodep->width()][idFrom(nodep)] = nodep;
+            m_removedTemps.emplace(nodep);
+        }
+
+        if (v3Global.opt.stats()) ++m_totalTemps;
+    }
+    void visit(AstVarRef* nodep) override {
+        if (!isWideTemp(nodep->varp()) || isExcluded(nodep->varp())) return;
+
+        if (nodep->access().isWriteOnly()) {
+            auto matchingCandidates = m_freeTemps.begin();
+            if (matchingCandidates == m_freeTemps.end() || matchingCandidates->second.empty()
+                || matchingCandidates->first < nodep->width()) {
+                // If we didn't find candidate with appropriate width, create one.
+                AstVar* const varp = createWideTemp(nodep->fileline(), nodep->dtypep());
+                const auto it
+                    = m_freeTemps.emplace(nodep->width(), IdTempMap{{idFrom(varp), varp}});
+                matchingCandidates = it.first;
+            }
+
+            UASSERT(matchingCandidates != m_freeTemps.end() && !matchingCandidates->second.empty(),
+                    "Candidate for substitution should be available or created previously");
+
+            const auto candidate = matchingCandidates->second.begin();
+            UINFO(5, "Replacing WRITE temp " << nodep->varp() << " with: " << candidate->second
+                                             << endl);
+
+            AstVar* const varp = candidate->second;
+
+            m_replaced.emplace(nodep->varp(), varp);
+            matchingCandidates->second.erase(candidate);
+            if (matchingCandidates->second.empty()) m_freeTemps.erase(matchingCandidates);
+
+            fixupType(nodep, varp);
```

Comment:
This seems like it picks the widest free temporary to replace an existing one. Does it help much to try to reuse wider temps in place of narrower ones? Could we just pick an available temporary with precisely the required width, and not reuse across mismatched widths? I think that would eliminate the need to create new temps here (you should never need more than you start with), so no risk of blowing up the design, and it would also eliminate the need for the implicit width conversion at runtime. I have a feeling that would cover a lot of reuse potential.